// node-os-utils/dist/test/unit/utils/platform-specific.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const platform_specific_1 = require("../../utils/platform-specific");
describe('CrossPlatformValidator.validateSystemInfo', () => {
    it('应该验证包含核心字段的macOS系统信息', () => {
        const info = {
            hostname: 'MacBook-Pro',
            platform: 'darwin',
            release: '23.4.0',
            version: 'macOS 14.4',
            arch: 'arm64',
            uptime: 3600000,
            uptimeSeconds: 3600,
            bootTime: Date.now() - 3600000,
            loadAverage: {
                load1: 1.2,
                load5: 0.8,
                load15: 0.5
            }
        };
        (0, chai_1.expect)(platform_specific_1.CrossPlatformValidator.validateSystemInfo(info)).to.be.true;
    });
    it('缺少必要字段时应该验证失败', () => {
        const invalidInfo = {
            platform: 'darwin',
            release: '23.4.0'
        };
        (0, chai_1.expect)(platform_specific_1.CrossPlatformValidator.validateSystemInfo(invalidInfo)).to.be.false;
    });
});
